### PR TITLE
[BEAM-12021] Add Pubsub overrides before portable proto translation in Dataflow

### PR DESCRIPTION
**Pull request by kennknowles**
These overrides were implicitly removed when we switched to using the original graph instead of the override graph. But Dataflow still needs these overrides.

Trying to take the key part of #14276 to fix Java postcommit.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by kennknowles** [comment]
run java postcommit

**Comment by kennknowles** [comment]
Run Java Examples on Dataflow Runner V2

**Comment by kennknowles** [comment]
CC: @tvalentyn

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java [hunk 1]
```diff
@@ -454,6 +454,17 @@ protected DataflowRunner(DataflowPipelineOptions options) {
     this.ptransformViewsWithNonDeterministicKeyCoders = new HashSet<>();
   }
 
+  /** For portable jobs, Dataflow still requires an override of the PubsubIO transforms. */
+  private List<PTransformOverride> getPortableOverrides() {
+    return ImmutableList.of(
+        PTransformOverride.of(
```
Comment (boyuanzz) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java:460]: We should also check `hasExperiment(options, "enable_custom_pubsub_source")`

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java [hunk 2]
```diff
@@ -454,6 +454,17 @@ protected DataflowRunner(DataflowPipelineOptions options) {
...
+            PTransformMatchers.classEqualTo(PubsubUnboundedSource.class),
+            new DataflowReadFromPubsubSourceForRunnerV2OverrideFactory()),
+        PTransformOverride.of(
```
Comment (boyuanzz) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java:463]: We should also check `hasExperiment(options, "enable_custom_pubsub_sink")`

**Review by boyuanzz — COMMENTED** [review]
I'm sorry that I may miss some content on why we move the `replace transform` step after generating fnapi proto. Would you like to elaborate more on that part?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java (hunk 1 above)
Comment (kennknowles) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java:460]: Yes, good catch.

**Review by kennknowles — COMMENTED** [review]
We move this to apply to the portable proto because that is what is sent to runner v2. So without applying this, runner v2 does not use the native Pubsub, yes?

**Comment by kennknowles** [comment]
The perfect situation would be to have zero overrides for runner v2 and do everything in the backend.

For now, we keep just the new Pubsub override. Hopefully we are able to get to zero.

**Comment by kennknowles** [comment]
You can see in https://ci-beam.apache.org/job/beam_PostCommit_Java_PR/631/ that this fixes the Pubsub postcommit. The other failure is the BigQuery storage test that is fixed in #14299

**Comment by kennknowles** [comment]
Since #14299 is merged, I will rebase and we will be sure to have up to date test signal. But you can also view the link above if you are interested.

**Comment by kennknowles** [comment]
run java postcommit

**Comment by kennknowles** [comment]
Run Java Examples on Dataflow Runner V2

**Comment by boyuanzz** [comment]
> We move this to apply to the portable proto because that is what is sent to runner v2. So without applying this, runner v2 does not use the native Pubsub, yes?

Correct. we might have situation like even with dataflow, some customers doesn't want to use Dataflow native PubSub.

**Comment by kennknowles** [comment]
Actually because of BEAM-7716 and BEAM-7717 probably noone should use the custom PubsubIO.

**Comment by kennknowles** [comment]
(but I am not trying to disable it in this PR, so I will check those experiments)

**Comment by kennknowles** [comment]
run java postcommit

**Comment by kennknowles** [comment]
Run Java Examples on Dataflow Runner V2

**Review by boyuanzz — APPROVED** [review]
Thanks. We can merge this PR once all tests pass.

**Comment by kennknowles** [comment]
Looks like this doesn't fix it: 

```
17:58:21 > Task :runners:google-cloud-dataflow-java:googleCloudPlatformRunnerV2IntegrationTest
18:40:01 
18:40:01 org.apache.beam.sdk.io.gcp.pubsub.PubsubReadIT > testReadPubsubMessageId FAILED
18:40:01     java.lang.AssertionError at PubsubReadIT.java:88
18:45:09 
18:45:09 org.apache.beam.sdk.io.gcp.pubsub.PubsubReadIT > testReadPublicData FAILED
18:45:09     java.lang.AssertionError at PubsubReadIT.java:60
19:37:34 
19:37:34 org.apache.beam.sdk.io.gcp.bigquery.BigQueryToTableIT > testNewTypesQueryWithoutReshuffle FAILED
19:37:34     java.lang.RuntimeException at BigQueryToTableIT.java:116
19:45:44 Build timed out (after 240 minutes). Marking the build as aborted.
```

The last successful build took 3 hours. I have not investigated whether there are new tests added to this job.

**Comment by kennknowles** [comment]
Clarification: the last successful build of the `_PR` job was 3 hours and a bit. The actual last successful postcommit I think was GC'd. The last one retained was 1 hour 45 minutes.

**Comment by kennknowles** [comment]
run java postcommit

**Comment by kennknowles** [comment]
Linking to most recent failure for archive before I adjust the PR: https://ci-beam.apache.org/job/beam_PostCommit_Java_PR/637/

**Comment by kennknowles** [comment]
run java postcommit

**Comment by kennknowles** [comment]
run java postcommit

**Comment by kennknowles** [comment]
Failure was not in Pubsub. This change really should work fine since we aren't even using portable job submission anymore.

**Comment by boyuanzz** [comment]
https://ci-beam.apache.org/job/beam_PostCommit_Java/7360/ is broken from head because of HealthCareIO quota issue.

**Comment by kennknowles** [comment]
OK. After #14321 I think it could make sense to even separate different ITs with different quotas. We might also need to put a retry limit or something on HealthCareIO or timeout on the specific test cases. It should fail instead of timing out so we can get a gradle scan, etc.

**Comment by kennknowles** [comment]
It seems like it does fail in a reasonable about of time on master so I am worried that we don't have enough signal for this PR.

**Comment by kennknowles** [comment]
re-running should hit Gradle cache for more things, perhaps that will get it under the deadline

**Comment by kennknowles** [comment]
run java postcommit

**Comment by kennknowles** [comment]
Re-reading this code, it is a bad cherrypick and edit. The change does not do what it says. It may have been useful for fixing Java Postcommit before PJS was rolled back, but now it doesn't matter.

**Comment by kennknowles** [comment]
On #14317 you make the good point that this may as well be merged there. So that is what I have done.

**Pull request closed without merging**